shared/dataflow/codeql/dataflow/internal/DataFlowImplCommon.qll

```diff
@@ -958,6 +961,377 @@ module MakeImplCommon<InputSig Lang> {
+     * call targets.
+     */
+    private predicate typeFlowArgTypeFromReturn(ArgNode arg, DataFlowType t) {
+      exists(ParamNode p, DataFlowType t1, DataFlowType t2 |
+        dataFlowTakenCallEdgeOut(_, _, arg, p) and
+        (if trackedParamType(p) then typeFlowParamType(p, t1, false) else nodeDataFlowType(p, t1)) and
+        nodeDataFlowType(arg, t2) and
+        if typeStrongerThan(t2, t1) then t = t2 else (compatibleTypes(t1, t2) and t = t1)
+      )
+    }
+
+    /**
+     * Holds if `t` is a possible type for the tracked argument `arg` in the call
+     * context `cc` and that the current data flow stage has reached this
+     * context.
+     */
+    private predicate typeFlowArgType(ArgNode arg, DataFlowType t, boolean cc) {
+      trackedArgType(arg) and
+      (
+        exists(ParamNode p, DataFlowType t1, DataFlowType t2 |
+          paramMustFlow(p, arg) and
+          typeFlowParamType(p, t1, cc) and
+          nodeDataFlowType(arg, t2) and
+          if typeStrongerThan(t2, t1) then t = t2 else (compatibleTypes(t1, t2) and t = t1)
+        )
+        or
+        cc = [true, false] and
+        not paramMustFlow(_, arg) and
+        nodeDataFlowType(arg, t)
+      )
+    }
+
+    predicate typeFlowStats(int nodes, int tuples) {
+      nodes =
+        count(Node n |
+          typeFlowParamType(n, _, _) or typeFlowArgTypeFromReturn(n, _) or typeFlowArgType(n, _, _)
+        ) and
+      tuples =
+        count(Node n, DataFlowType t, boolean cc |
+          typeFlowParamType(n, t, cc)
+          or
+          typeFlowArgTypeFromReturn(n, t) and cc = false
+          or
+          typeFlowArgType(n, t, cc)
+        )
+    }
+
+    /**
+     * Holds if the `arg`-to-`p` edge should be considered for validation of the
+     * corresponding call edge in the in-going direction.
+     */
+    private predicate relevantArgParamIn(ArgNode arg, ParamNode p, DataFlowType pt) {
+      exists(DataFlowCall call, DataFlowCallable c |
+        Input::relevantCallEdgeIn(call, c) and
+        callEdge(call, c, arg, p) and
+        paramMustFlow(_, arg) and
+        trackedArgType(arg) and
+        nodeDataFlowType(p, pt)
+      )
+    }
+
+    /**
+     * Holds if there is a possible type for `arg` in the call context `cc` that
+     * is consistent with the static type of `p`.
+     */
+    private predicate validArgParamIn(ArgNode arg, ParamNode p, boolean cc) {
+      exists(DataFlowType t1, DataFlowType t2 |
+        typeFlowArgType(arg, t1, cc) and
+        relevantArgParamIn(arg, p, t2) and
+        compatibleTypes(t1, t2)
+      )
+    }
+
+    /**
+     * Holds if the edge `call`-to-`c` is valid in the in-going direction in the
+     * call context `cc`.
+     */
+    predicate typeFlowValidEdgeIn(DataFlowCall call, DataFlowCallable c, boolean cc) {
+      Input::relevantCallEdgeIn(call, c) and
+      cc = [true, false] and
+      (
+        not Input::enableTypeFlow()
+        or
+        forall(ArgNode arg, ParamNode p |
```

Comment:
Recursion through `forall` 😨

Reply:
Just to understand: `forall` is needed because there can be either no `arg`-`p` pairs, or possibly multiple pairs that we all want to verify?

Reply:
Indeed, this is what's needed. We want to verify all of them. And yes, it's also possible that there can be zero `arg`-`p` pairs to verify.